playbooks/nvidia-mig.yml

```diff
@@ -5,96 +5,94 @@
+      tags: always
 
-    # Manage MIG
-    - name: enable MIG mode
-      command: nvidia-smi -mig 1
-      tags: enable, never
-    - name: disable MIG mode
-      command: nvidia-smi -mig 0
-      tags: disable, never
+    # Fail if MIG is requested to be enabled, but no GPUs support MIG. The
+    # opposite is considered ok as a GPU with no MIG support can be considered
+    # having it always disabled.
+    - fail:
+        msg: "No GPUs supporting MIG found while MIG was requested to be enabled."
+      when: mig_enabled_count == "0" and mig_disabled_count == "0"
+      tags: enable
 
-    # Post-tasks
-    - name: wait for MIG stuff to settle down and nvidia-persistenced to start again
-      pause:
-        seconds: 20
-      tags: enable, disable, never
-    - name: stop system services
-      systemd:
-        state: stopped
-        enabled: no
-        name: "{{ item }}"
-      with_items: "{{ nv_services }}"
-      tags: enable, disable, never
-    - name: unload drivers
-      modprobe:
-        state: absent
-        name: "{{ item }}"
-      with_items: "{{ nv_modules }}"
-      tags: enable, disable, never
-    - name: start fabric manager
-      systemd:
-        state: started
-        name: nvidia-fabricmanager
-      tags: enable, disable, never
-    - name: stop nvidia-persistenced again
-      systemd:
-        state: stopped
-        name: nvidia-persistenced
-      tags: enable, disable, never
-    - name: reset GPUs
-      command: nvidia-smi --gpu-reset
-      tags: enable, disable, never
-    - name: load drivers
-      modprobe:
-        state: present
-        name: "{{ item }}"
-      with_items: "{{ nv_modules }}"
-      ignore_errors: true
-      tags: enable, disable, never
-    - name: start system services
-      systemd:
-        state: started
-        enabled: yes
-        name: "{{ item }}"
-      with_items: "{{ nv_services }}"
-      ignore_errors: true
+    - name: Check that we need to enable MIG on any GPUs
+      set_fact:
+        need_changes: True
+      when: mig_disabled_count != "0"
+      tags: enable
+
+    - name: Check that we need to disable MIG on any GPUs
+      set_fact:
+        need_changes: True
+      when: mig_enabled_count != "0"
+      tags: disable
+
+    - name: Display info if no MIG changes are needed
+      debug:
+        msg: "All GPUs in the desired MIG state already, skipping most of the other tasks."
+      when: not need_changes
       tags: enable, disable, never
 
+    - name: Apply new MIG mode settings
+      block:
+        # Pre-tasks
+        - name: collect facts about services
+          service_facts:
+          tags: enable, disable, never
+
+        - name: stop system services referencing nvidia
+          systemd:
+            state: stopped
+            name: "{{ item }}"
+          loop: "{{ nv_services }}"
+          when: item in ansible_facts.services and ansible_facts.services[item].state == "running"
```

Comment:
nvsm needs to be stopped, but nvsm does not show up in state `running` as the systemd service launches several other services.

Actual status of an active nvsm shown in systemd:

```
# systemctl status nvsm
● nvsm.service - NVIDIA System Management service suite
   Loaded: loaded (/usr/lib/systemd/system/nvsm.service; enabled; vendor preset: enabled)
   Active: active (exited) since Mon 2020-08-24 12:46:28 PDT; 11s ago
  Process: 15685 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 15685 (code=exited, status=0/SUCCESS)
```

Status as shown from Ansible `service_facts`:

```
            "nvsm.service": {
                "name": "nvsm.service",
                "source": "systemd",
                "state": "stopped",
                "status": "enabled"
            },
```

When I ran this playbook, it skipped stopping nvsm:
```
TASK [restart system services referencing nvidia] ******************************************************************************************************************************************************************
changed: [myhostname] => (item=dcgm.service)
skipping: [myhostname] => (item=nvsm.service)
changed: [myhostname] => (item=docker.service)
```

Reply:
Thanks, let me see about nvsm, stopping unconditionally and restarting seems pretty hacky. How many sub-daemons is it running?

Reply:
@pjaroszynski: This is the list of NVSM services I see running on my test host.

```
# systemctl | grep nvsm
  nvsm-api-gateway.service
       loaded active running   NVSM API Server to provide DGX System Management APIs

  nvsm-mqtt.service
       loaded active running   MQTT broker for NVSM API for signaling within NVSM API components

  nvsm-notifier.service
       loaded active running   NVSM Notifier service.

  nvsm-plugin-environment.service
       loaded active running   NVSM API plugin service to monitor changes in environmental sensors

  nvsm-plugin-gpumonitor.service
       loaded active running   NVSM API plugin service to monitor GPU events from DCGM

  nvsm-plugin-memory.service
       loaded active running   NVSM API plugin service to monitor system memory devices

  nvsm-plugin-monitor-environment.service
       loaded active running   Environmental DSHM service.

  nvsm-plugin-monitor-network.service
       loaded active running   Switch DSHM service.

  nvsm-plugin-monitor-pcie.service
       loaded active running   System DSHM service.

  nvsm-plugin-monitor-storage.service
       loaded active running   Storage DSHM service.

  nvsm-plugin-monitor-system.service
       loaded active running   System DSHM service.

  nvsm-plugin-network.service
       loaded active running   NVSM API plugin service to monitor changes in network

  nvsm-plugin-pcie.service
       loaded active running   NVSM API plugin service to monitor system PCIe devices

  nvsm-plugin-processor.service
       loaded active running   NVSM API plugin service to monitor system processors

  nvsm-plugin-storage.service
       loaded active running   NVSM API plugin service to monitor System Storage

  nvsm-selwatcher.service
       loaded active running   NVSM API plugin service to monitor SEL events from BMC

  nvsm.service
       loaded active exited    NVIDIA System Management service suite
```